2.7 session concurrency

AD7six opened the pull request:
This starts with a failing test, fixes the test and then adjusts the options to reduce the code execution time.

We _could_ rework the write method to be of the form:

```
$db = x
$db->update($this->_model, ...) || $db->create($this->_model, ...)
```

But that's maybe not really necessary (or rather, not worth the maintenance of effectively re-implementing the model save method)

Fixes #2808

AD7six committed:
Add a failing test for session concurrency problems

AD7six committed:
Ignore counter cache also

markstory commented on lib/Cake/Model/Datasource/Session/DatabaseSession.php:
```diff
@@ -114,7 +114,17 @@ public function write($id, $data) {
+		try {
+			return $this->_model->save($record, $options);
+		} catch (Exception $e) {
+			return $this->_model->save($record, $options);
```
What happens if it fails again?

AD7six replied:
If it fails again, it's because of a none-transient problem and that'll bubble up (rightly so IMO).

I see I left it catching all exceptions, I'll update that to PDOException since that's the only scenario that this code is expecting to encounter.

markstory commented:
Other than my one question this looks like an improvement. Disabling code we don't need will help shorten the race window.

AD7six committed:
If a PDOException is thrown, try again.

This can occur if the exists check returns no rows, but before this
request inserts into the database a _different_ thread inserts a
session record.

AD7six committed:
Remove some exists checks by not validating

AD7six committed:
Remove callbacks to speed up saving

AD7six committed:
Skip Sqlite

Sqlite will accept an insert with duplicate primary keys and silently
drop the query. There is no exception, or indication that the insert
failed.

markstory commented:
Should this also be backported to master? It is probably also an issue in 2.6.x

AD7six commented:
@markstory the problem certainly exists in master - does the different save calls have any BC concerns? If not, I'd say yes it should be.

markstory commented:
I don't think the BC concern out weighs the issue around concurrency problems.
